L42-python-metadata-flags.md

```diff
@@ -0,0 +1,153 @@
+
+### The 3-states switch of 'wait-for-ready' in C-Core
+
+The implementation of 'wait-for-ready' flags is not as simple as a one bit switch. It contains two bits one is `GRPC_INITIAL_METADATA_WAIT_FOR_READY` another one is `GRPC_INITIAL_METADATA_WAIT_FOR_READY_EXPLICITLY_SET`. Although two bits can represent four states, three of them are valid.
+
+Assume `flag` is the flag variable we are manipulating.
+
+If the developer wants to use default value, then `flag = flag & ~WAIT_FOR_READY & ~WAIT_FOR_READY_EXPLICITLY_SET` (both 0).
+
+If the developer doesn't want to 'wait-for-ready', then `flag = flag & ~WAIT_FOR_READY & WAIT_FOR_READY_EXPLICITLY_SET` (0, 1).
+
+If the developer wants to 'wait-for-ready', then `flag = flag & WAIT_FOR_READY & WAIT_FOR_READY_EXPLICITLY_SET` (1, 1).
+
+### Related Proposals: 
+
+N/A
+
+## Proposal
+
+* Add an optional `wait_for_ready` variable to `MultiCallable` classes initialization methods. Default `None`, accept `bool`.
+* Import initial metadata flags constants from `grpc_types.h` to `grpc.pxi`.
+
+### DEMO snippets 
+
+```Python
+# Per RPC level
+stub = ...Stub(...)
+
+stub.initial_rpc_call(..., wait_for_ready=True)
+stub.important_transaction_1(..., wait_for_ready=True)
+stub.unimportant_transaction_2(...)
+stub.important_transaction_3(..., wait_for_ready=True)
+stub.unimportant_transaction_4(...)
+# The unimportant transactions can be status report, or health check, etc.
+```
+
+```Python
+# Per RPC level with configuration for subsets
+stub = ...Stub(...)
+
+important_call_options = {'wait_for_ready': True, 'timeout': 4000}
+failfast_call_options = {'wait_for_ready': False}
+
+stub.initial_rpc_call(..., **important_call_options)
+stub.important_transaction_1(..., **important_call_options)
+stub.unimportant_transaction_2(..., **failfast_call_options)
+stub.important_transaction_3(..., **important_call_options)
+stub.unimportant_transaction_4(..., **failfast_call_options)
+```
+
+### Unsupported Design
+
+```Python
+# Channel level - Invalid Design
+channel = grpc.insecure_channel(..., wait_for_ready=True)
+stub = ...Stub(channel)
+
+stub.important_transaction_1(...)
+stub.important_transaction_2(...)
+stub.unimportant_transaction_3(...) # This will wait as well; user might get confused
+# --- A lot of code here ---
+stub.yet_another_transaction_500(...) # User might forget the channel setting
+```
+
+## Rationale
+
+### Default `None`, accept `bool` value
+* The `None` state defined that this value haven't been touched
+* It might be useful in some scenarios that interceptors wants to read its value and make changes
+
+### Whether to add channel level variable or not
+* The channel level variable is convenient and Pythonic
+* But it may create extra burden for future API evolution
+* **Please** comment about whether you feel like this design works for you or not
```

Comment:
Before merging, I think it would be helpful to change this line to something like "Comments supporting the decision not to support channel-level configuration:"